beacon_node/eth1/src/service.rs

```diff
@@ -34,6 +35,223 @@ const GET_DEPOSIT_LOG_TIMEOUT_MILLIS: u64 = STANDARD_TIMEOUT_MILLIS;
 
 const WARNING_MSG: &str = "BLOCK PROPOSALS WILL FAIL WITHOUT VALID ETH1 CONNECTION";
 
+/// Applies the given function $func_name to all usable endpoints of $endpoints until
+/// one of them returns Ok or if no usable endpoint is left it returns the error of the last usable
+/// endpoint. If no endpoint is usable it returns None. Usability of endpoints is checked lazily
+/// using the EndpointsCache structure. For each endpoint if it returns an error the on_err function
+/// is called.
+macro_rules! with_fallback_and_on_err {
```

Comment:
The `macro` is cool, but I'm not convinced that we need to use a macro here. Generally, I find *not* writing a macro is the more maintainable and extensible solution; compile errors are simpler, imports/exports are easier and the types/generics are more clearly expressed. Of course, there are times where macros are needed or are simpler.

This is how I would go about this problem without templating/macros: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=564a7fc931b10cfc757f752afe261604

Considering that we're going to need this same fallback logic in the VC, I would suggest to use a struct-based approach (like in the the play above) defined in a new crate in `common/` that can be shared between `validator_client/` and `beacon_node/eth1`.

Reply:
You can even include the validator client fallback in this PR, if you like like :)

Reply:
thanks for the detailed review. I wanted to go a similar route at the beginning but then probably abandoned the idea to early after having problems integrating everything in an async way. I will have a try at that again...